components/validator/l2_output_submitter.go

```diff
@@ -37,33 +57,131 @@ func NewL2OutputSubmitter(cfg Config, l log.Logger) (*L2OutputSubmitter, error)
-// It returns: the next block number, if the output should be made, error
-func (l *L2OutputSubmitter) FetchNextOutputInfo(ctx context.Context) (*eth.OutputResponse, bool, error) {
-	nextBlockCtx, cancelNextBlockCtx := context.WithTimeout(ctx, l.cfg.NetworkTimeout)
-	defer cancelNextBlockCtx()
-	callOpts := utils.NewCallOptsWithSender(nextBlockCtx, l.cfg.TxManager.From())
-	nextCheckpointBlock, err := l.l2ooContract.NextBlockNumber(callOpts)
+func (l *L2OutputSubmitter) Start() error {
+	l.log.Info("starting L2 Output Submitter")
+	l.wg.Add(1)
+	go l.loop()
+
+	return nil
+}
+
+func (l *L2OutputSubmitter) Stop() {
+	l.log.Info("stopping L2 Output Submitter")
+	l.cancel()
+	l.wg.Wait()
+}
+
+func (l *L2OutputSubmitter) loop() {
+	defer l.wg.Done()
+
+	l.submitChan <- struct{}{}
+
+	for {
+		select {
+		case <-l.submitChan:
+			if err := l.trySubmitL2Output(); err != nil {
+				l.log.Error("failed to submit l2 output", "err", err)
+			}
+		case <-l.ctx.Done():
+			return
+		}
+	}
+}
+
+func (l *L2OutputSubmitter) trySubmitL2Output() error {
+	nextBlockNumber, canSubmit, err := l.canSubmit(l.ctx)
+	if err != nil {
+		return fmt.Errorf("failed to check if it can submit: %w", err)
+	}
+	if !canSubmit {
+		return nil
+	}
+
+	output, canSubmit, err := l.fetchOutput(l.ctx, nextBlockNumber)
+	if err != nil {
+		return fmt.Errorf("failed to fetch next output: %w", err)
+	}
+	if !canSubmit {
+		return nil
+	}
+
+	data, err := submitL2OutputTxData(l.l2ooABI, output, l.cfg.OutputSubmitterBondAmount)
+	if err != nil {
+		return fmt.Errorf("failed to create submit l2 output transaction data: %w", err)
+	}
+
+	l.submitL2OutputTx(data)
+
+	l.metr.RecordL2OutputSubmitted(output.BlockRef)
+
+	return nil
+}
+
+// canSubmit checks if submission interval has elapsed and selected for next validator.
+func (l *L2OutputSubmitter) canSubmit(ctx context.Context) (*big.Int, bool, error) {
+	currentBlockNumber, nextBlockNumber, err := l.fetchBlockNumbers(ctx)
 	if err != nil {
-		l.log.Error("validator unable to get next block number", "err", err)
 		return nil, false, err
 	}
 
+	var nextBlockNumberToWait *big.Int
+	if l.cfg.RollupConfig.IsBlueBlock(nextBlockNumber.Uint64()) {
+		nextBlockNumberToWait = new(big.Int).Add(nextBlockNumber, common.Big1)
+	} else {
+		nextBlockNumberToWait = nextBlockNumber
+	}
+
+	// Wait for L2 blocks proceeding when validator submission interval has not elapsed
+	if currentBlockNumber.Cmp(nextBlockNumberToWait) < 0 {
+		return nil, false, l.waitL2Blocks(ctx, currentBlockNumber, nextBlockNumberToWait)
+	}
+
+	// Check if selected for next validator or not
+	isNextValidator, err := l.checkNextValidator(ctx)
+	if !isNextValidator {
+		return nil, false, nil
+	}
+
+	return nextBlockNumber, true, nil
+}
+
+func (l *L2OutputSubmitter) fetchBlockNumbers(ctx context.Context) (*big.Int, *big.Int, error) {
+	cCtx, cancel := context.WithTimeout(ctx, l.cfg.NetworkTimeout)
+	callOpts := utils.NewCallOptsWithSender(cCtx, l.cfg.TxManager.From())
```

Comment:
Instead of declaring `callOpts` in each function, why don't we reuse it with declaring in `L2OutputSubmitter`, as done in [`Challenger`](https://github.com/kroma-network/kroma/blob/9b2e8984db0b74dd267aa87119bab4b92354cc6c/components/validator/challenger.go#L52)?

Reply:
As explained offline, it's because we're using `callOpts` with network timeout context. We need to declare the context for each network call.